### PR TITLE
Add more metadata to BuildLog Telemetry

### PR DESCRIPTION
**Pull request by fkpwolf**
[AB#397541](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/397541)

### CODE REVIEW

File: src/docfx/lib/log/Telemetry.cs [hunk 1]
```diff
@@ -151,12 +151,24 @@ public static void TrackExternalLink(string tag, string attribute, string scheme
...
+                error.MsTechnology,
+                error.MsService,
+                error.MsSubservice,
+            };
+            var additoinInfoString = JsonUtility.Serialize(additionalInfo);
             if (!s_isRealTimeBuild.Value)
             {
                 s_errorCountMetric.TrackValue(
-                    1, code, level.ToString(), CoalesceEmpty(name), "User", s_os, s_version, s_repo, s_branch, s_correlationId, s_sessionId);
+                    1, code, level.ToString(), CoalesceEmpty(name), additoinInfoString, s_os, s_version, s_repo, s_branch, s_correlationId, s_sessionId);
```
Comment (hejingkan2005) [inline src/docfx/lib/log/Telemetry.cs]: Is it okay to remove type = "User" here?

Comment (hejingkan2005) [reply src/docfx/lib/log/Telemetry.cs]: Or we can include 'User' into above additionalInfo object? If yes, I think we need to keep its name as it is.

**Review by hejingkan2005 — APPROVED** [review]

File: src/docfx/lib/log/Telemetry.cs (hunk 1 above)
Comment (hejingkan2005) [reply src/docfx/lib/log/Telemetry.cs]: Offline talked to @fkpwolf , he has confirmed with @yufeih that 'User' is hard-coded, and not used. So we can discard it.

File: src/docfx/lib/log/ErrorLog.cs [hunk 2]
```diff
@@ -40,10 +40,13 @@ public override void Add(Error error)
...
-                    if (msAuthor != default)
-                    {
-                        error = error with { MsAuthor = msAuthor };
-                    }
+                    var msProd = MetadataProvider?.GetMetadata(Null, source).MsProd;
+                    var msTechnology = MetadataProvider?.GetMetadata(Null, source).MsTechnology;
+                    var msService = MetadataProvider?.GetMetadata(Null, source).MsService;
+                    var msSubservice = MetadataProvider?.GetMetadata(Null, source).MsSubservice;
+
+                    error = error with
```
Comment (herohua) [inline src/docfx/lib/log/ErrorLog.cs:48]: would default overwrite existing value?

File: src/docfx/lib/log/ErrorLog.cs [hunk 3]
```diff
@@ -40,10 +40,13 @@ public override void Add(Error error)
                 try
                 {
                     var msAuthor = MetadataProvider?.GetMetadata(Null, source).MsAuthor;
-                    if (msAuthor != default)
-                    {
-                        error = error with { MsAuthor = msAuthor };
-                    }
+                    var msProd = MetadataProvider?.GetMetadata(Null, source).MsProd;
```
Comment (yufeih) [inline src/docfx/lib/log/ErrorLog.cs:43]: call `GetMetadata` once.

File: src/docfx/lib/log/Error.cs [hunk 4]
```diff
@@ -27,6 +25,16 @@ internal record Error
 
         public object?[] MessageArguments { get; init; } = Array.Empty<object?>();
 
+        public string? MsAuthor { get; init; }
```
Comment (yufeih) [inline src/docfx/lib/log/Error.cs:28]: Consider extract a class `AdditonalErrorInfo` and put these `Ms.xyz` into that class.

File: src/docfx/lib/log/Telemetry.cs [hunk 5]
```diff
@@ -151,12 +151,24 @@ public static void TrackExternalLink(string tag, string attribute, string scheme
...
+            var name = error.PropertyPath;
+            var additionalInfo = new
+            {
+                error.MsAuthor,
+                error.MsProd,
+                error.MsTechnology,
+                error.MsService,
+                error.MsSubservice,
+            };
+            var additionalInfoString = JsonUtility.Serialize(additionalInfo);
```
Comment (yufeih) [inline src/docfx/lib/log/Telemetry.cs:167]: Consider use `ms.author` or `ms_author` as key in the output JSON. As a guideline, try to use the same name and case for the same term, if `.` is not allowed, replace it with `_`.

**Review by yufeih — APPROVED** [review]

**Comment by fkpwolf** [comment]
the diff are excepted: now log have more properties like `ms.prod`.  So this PR is ready for merge @yufeih

**Review by live1206 — APPROVED** [review]

**Pull request merged**